spring-boot/src/main/java/org/booking/restcontroller/auth/AuthRestController.java

```diff
@@ -0,0 +1,94 @@
+package org.booking.restcontroller.auth;
+
+import org.booking.auth.JwtUtil;
+import org.booking.model.AccessToken;
+import org.booking.model.UserInfo;
+import org.booking.restcontroller.dto.BookingResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/auth")
+public class AuthRestController {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AuthRestController.class);
+
+    @Value("${spring.security.oauth2.client.provider.github.token-uri}")
+    private String accessTokenUrl;
+
+    @Value("${spring.security.oauth2.client.provider.github.user-info-uri}")
+    private String userInfoUrl;
+
+    @Value("${spring.security.oauth2.client.registration.github.client-id}")
+    private String clientId;
+
+    @Value("${spring.security.oauth2.client.registration.github.client-secret}")
+    private String clientSecret;
+
+    @Value("${spring.security.oauth2.client.registration.github.redirect-uri}")
+    private String redirectUri;
+
+    private final RestTemplate restTemplate;
+
+    public AuthRestController(RestTemplate restTemplate) {
+        this.restTemplate = restTemplate;
+    }
+
+    @GetMapping("/callback")
+    public ResponseEntity<BookingResponse> callback(@RequestParam(name = "code") String code) {
+        return getAccessToken(code)
+                .flatMap(this::createAuthToken)
+                .map(JwtUtil::createToken)
+                .map(token -> ResponseEntity.ok(new BookingResponse(token)))
```

Comment:
Implemented GitHub OAuth2 login flow: Users are redirected to GitHub for authentication, then the app exchanges the authorization code for an access token. The token is used to retrieve user information, which is then processed to generate a JWT for application access.